### PR TITLE
Handle Piecewise in TransformVisitor (for e.g. cse)

### PR DESCRIPTION
**Pull request by bjodah**
This should enable cse to also pierce into `Piecewise` instances.
Fixed gh-1952.

Thank you @isuruf for the pointer!

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,7 @@
 cmake_minimum_required(VERSION 2.8.12)
-
+if (POLICY CMP0074)
+  cmake_policy(SET CMP0074 NEW)
```
Comment (bjodah) [inline CMakeLists.txt:3]: This allows me to set the environment variable `FLINT_ROOT` for example. I think this is rather harmless to enable, no?

File: symengine/visitor.cpp [hunk 2]
```diff
@@ -201,6 +201,31 @@ void TransformVisitor::bvisit(const MultiArgFunction &x)
...
+    auto branch_cond_pairs = x.get_vec();
+    PiecewiseVec new_pairs;
+    bool changed = false;
+    for (const auto &branch_cond : branch_cond_pairs) {
+        auto branch = branch_cond.first;
+        auto cond = branch_cond.second;
+        auto new_branch = apply(branch);
+        // decltype(cond) new_cond =
+        // rcp_static_cast<some-kind-of-boolean-cse-symbol?>(apply(rcp_static_cast<const
+        // Basic>(cond)));
```
Comment (bjodah) [inline symengine/visitor.cpp:215]: I couldn't figure out how to do subexpression elimination on the conditionals, so before merging I should probably just delete these comments...

File: symengine/tests/basic/test_cse.cpp [hunk 3]
```diff
@@ -1,19 +1,25 @@
 #include "catch.hpp"
+#include "symengine/dict.h"
+#include "symengine/logic.h"
```
Comment (isuruf) [inline symengine/tests/basic/test_cse.cpp:3]: Duplicate

File: symengine/tests/basic/test_cse.cpp [hunk 4]
```diff
@@ -1,19 +1,25 @@
 #include "catch.hpp"
+#include "symengine/dict.h"
```
Comment (isuruf) [inline symengine/tests/basic/test_cse.cpp:2]: Use `<` instead of `"`

Comment (bjodah) [reply symengine/tests/basic/test_cse.cpp]: thanks, my IDE seem to have "helped" me here... dropping those includes.

**Comment by isuruf** [comment]
TransformVisitor seems to be missing other classes as well. Can you open an issue?

File: symengine/tests/basic/test_cse.cpp [hunk 5]
```diff
@@ -302,6 +306,30 @@ TEST_CASE("CSE: simple", "[cse]")
...
+        vec_pair substs;
+        vec_basic reduced;
+        cse(substs, reduced, {pw1});
+        REQUIRE(unified_eq(substs, {{x0, add(x, y)}}));
+        REQUIRE(unified_eq(reduced, {piecewise({{pow(x0, i2), Gt(x, y)},
+                                                {sqrt(x0), boolTrue}})}));
+    }
+    {
+        auto pw2 = piecewise({{pow(x, i2), Gt(add(x, y), i2)},
+                              {sqrt(y), Gt(add(x, y), i3)},
```
Comment (isuruf) [inline symengine/tests/basic/test_cse.cpp:322]: Can you swap the two conditions? They don't make much sense.

Comment (bjodah) [reply symengine/tests/basic/test_cse.cpp:322]: Sorry about that, fixed in 3570b305a835f3d57d71d98c6d6ed3ee87863eb0

**Comment by bjodah** [comment]
> TransformVisitor seems to be missing other classes as well. Can you open an issue?

Sure, but it's not obvious to me what other classes you are referring to?

File: symengine/tests/basic/test_cse.cpp (hunk 5 above)
Comment (isuruf) [reply symengine/tests/basic/test_cse.cpp:322]: Looks like you only changed part of the test

**Comment by isuruf** [comment]
> Sure, but it's not obvious to me what other classes you are referring to?

I didn't do a thorough check, but found classes like `Contains` are not implemented.

**Pull request merged**